File: eaze-referral-service/app/routers/redirect.py
from fastapi import APIRouter, Depends, Request
from fastapi.responses import RedirectResponse
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.db import get_db
from app.models import ReferralClick, ReferralCode
from app.services.db_helpers import transaction

router = APIRouter(tags=["redirect"])


@router.get("/r/{code}")
async def click_redirect(code: str, request: Request, session: AsyncSession = Depends(get_db)) -> RedirectResponse:
    """The link embedded in the copyable share message. Logs the click (funnel visibility only —
    never a source of reward attribution, see referral_service.signup_match) then forwards to the
    right store based on user agent."""
    settings = get_settings()

    result = await session.execute(
        select(ReferralCode.id).where(ReferralCode.code == code, ReferralCode.is_active.is_(True))
    )
    code_id = result.scalar_one_or_none()

    if code_id is None:
        return RedirectResponse(settings.fallback_url, status_code=302)

    user_agent = request.headers.get("user-agent", "")
    async with transaction(session):
        session.add(
            ReferralClick(
                referral_code_id=code_id,
                ip_address=request.client.host if request.client else None,
                user_agent=user_agent or None,
            )
        )

    ua = user_agent.lower()
    if "android" in ua:
        target = settings.play_store_url
    elif "iphone" in ua or "ipad" in ua:
        target = settings.app_store_url
    else:
        target = settings.fallback_url

    # Pass the code through as `referrer` too, so a future Play Install Referrer integration
    # (REFERRAL_PROGRAM_PLAN.md §3) can read it straight off the install without extra wiring.
    separator = "&" if "?" in target else "?"
    return RedirectResponse(f"{target}{separator}referrer={code}", status_code=302)

File: eaze-referral-backend/src/routes/redirect.ts
import { Router } from 'express';
import { pool } from '../db';
import { asyncHandler } from '../lib/asyncHandler';

export const redirectRouter = Router();

const PLAY_STORE_URL = process.env.PLAY_STORE_URL ?? 'https://play.google.com/store/apps/details?id=com.eaze.app';
const APP_STORE_URL = process.env.APP_STORE_URL ?? 'https://apps.apple.com/app/idXXXXXXXXX';
const FALLBACK_URL = process.env.FALLBACK_URL ?? PLAY_STORE_URL;

// GET /r/:code — the link embedded in the copyable share message. Logs the click (this is the
// "tracking to verify users who used the link" piece) then forwards to the right store.
redirectRouter.get('/r/:code', asyncHandler(async (req, res) => {
  const codeRow = await pool.query<{ id: number }>(
    'SELECT id FROM referral_codes WHERE code = $1 AND is_active = TRUE',
    [req.params.code]
  );

  if (!codeRow.rows[0]) {
    res.redirect(302, FALLBACK_URL);
    return;
  }

  await pool.query(
    'INSERT INTO referral_clicks (referral_code_id, ip_address, user_agent) VALUES ($1, $2, $3)',
    [codeRow.rows[0].id, req.ip, req.get('user-agent') ?? null]
  );

  const ua = (req.get('user-agent') ?? '').toLowerCase();
  const target = ua.includes('android') ? PLAY_STORE_URL : ua.includes('iphone') || ua.includes('ipad') ? APP_STORE_URL : FALLBACK_URL;

  // Pass the code through as `referrer` too, so a future Play Install Referrer integration
  // (REFERRAL_PROGRAM_PLAN.md §3) can read it straight off the install without extra wiring.
  const separator = target.includes('?') ? '&' : '?';
  res.redirect(302, `${target}${separator}referrer=${encodeURIComponent(req.params.code)}`);
}));
